edx_event_bus_kafka/publishing/event_producer.py

```diff
@@ -123,10 +125,23 @@ def get_serializer(signal: OpenEdxPublicSignal) -> AvroSignalSerializer:
 # fall out of scope and be garbage-collected, destroying the
 # outbound-message queue and threads. The use of this cache allows the
 # producers to be long-lived.
+#
+# We are also likely to need to iterate through this cache at server
+# shutdown in order to flush each of the producers, which means the
+# cache needs to never evict. See https://github.com/openedx/event-bus-kafka/issues/11
+# for more details.
+#
+# (Why not change the code to use a single Producer rather than multiple
+# SerializerProducer? Because the code actually turns out to be significantly
+# uglier that way due to the number of separate values that need to be passed
+# around in bundles. There aren't clear "cut-off" points. Additionally, it
+# makes unit testing harder/uglier since now the mocks need to either deal with
+# serialized bytes or mock out the serializers. Getting this down to a single
+# Producer doesn't really seem worth the trouble.)
```

Comment:
Before making this PR, I first tried to implement the single-producer approach. It would _work_, but it's ugly:  https://github.com/openedx/event-bus-kafka/compare/timmc/fix-settings...timmc/single-producer